# Mnist Dataseet
The MNIST dataset is a widely used dataset in the field of machine learning and computer vision. It consists of 70,000 grayscale images of handwritten digits (0-9), with each image being 28x28 pixels in size. The dataset is commonly used for training and testing various machine learning algorithms, particularly in the context of image classification tasks.

In [2]:
import pandas as pd
import numpy as np

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import (
    accuracy_score,
    classification_report,
    confusion_matrix
)

from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import LinearSVC
from sklearn.naive_bayes import MultinomialNB

### Loading the dataset

In [4]:
# Loading the Data 
df = pd.read_csv('train_for_mnist.csv')
df.head()

,label,pixel0,pixel1,pixel2,pixel3,pixel4,pixel5,pixel6,pixel7,pixel8,...,pixel774,pixel775,pixel776,pixel777,pixel778,pixel779,pixel780,pixel781,pixel782,pixel783
0,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
1,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
2,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
3,4,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0


### Describing the dataset

In [5]:
# Shape (rows, columns)
print(f"Dataset has {df.shape[0]} rows and {df.shape[1]} columns.")

# Column names
print("\nColumns:", df.columns.tolist())

# Data types
print("\nData types:")
print(df.dtypes)

# Summary statistics
df.describe()

Dataset has 42000 rows and 785 columns.

Columns: ['label', 'pixel0', 'pixel1', 'pixel2', 'pixel3', 'pixel4', 'pixel5', 'pixel6', 'pixel7', 'pixel8', 'pixel9', 'pixel10', 'pixel11', 'pixel12', 'pixel13', 'pixel14', 'pixel15', 'pixel16', 'pixel17', 'pixel18', 'pixel19', 'pixel20', 'pixel21', 'pixel22', 'pixel23', 'pixel24', 'pixel25', 'pixel26', 'pixel27', 'pixel28', 'pixel29', 'pixel30', 'pixel31', 'pixel32', 'pixel33', 'pixel34', 'pixel35', 'pixel36', 'pixel37', 'pixel38', 'pixel39', 'pixel40', 'pixel41', 'pixel42', 'pixel43', 'pixel44', 'pixel45', 'pixel46', 'pixel47', 'pixel48', 'pixel49', 'pixel50', 'pixel51', 'pixel52', 'pixel53', 'pixel54', 'pixel55', 'pixel56', 'pixel57', 'pixel58', 'pixel59', 'pixel60', 'pixel61', 'pixel62', 'pixel63', 'pixel64', 'pixel65', 'pixel66', 'pixel67', 'pixel68', 'pixel69', 'pixel70', 'pixel71', 'pixel72', 'pixel73', 'pixel74', 'pixel75', 'pixel76', 'pixel77', 'pixel78', 'pixel79', 'pixel80', 'pixel81', 'pixel82', 'pixel83', 'pixel84', 'pixel85', 'pix

,label,pixel0,pixel1,pixel2,pixel3,pixel4,pixel5,pixel6,pixel7,pixel8,...,pixel774,pixel775,pixel776,pixel777,pixel778,pixel779,pixel780,pixel781,pixel782,pixel783
count,42000.000000,42000.0,42000.0,42000.0,42000.0,42000.0,42000.0,42000.0,42000.0,42000.0,...,42000.000000,42000.000000,42000.000000,42000.00000,42000.000000,42000.000000,42000.0,42000.0,42000.0,42000.0
mean,4.456643,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.219286,0.117095,0.059024,0.02019,0.017238,0.002857,0.0,0.0,0.0,0.0
std,2.887730,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,6.312890,4.633819,3.274488,1.75987,1.894498,0.414264,0.0,0.0,0.0,0.0
min,0.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.0,0.0,0.0,0.0
25%,2.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.0,0.0,0.0,0.0
50%,4.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.0,0.0,0.0,0.0
75%,7.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.000000,0.000000,0.000000,0.00000,0.000000,0.000000,0.0,0.0,0.0,0.0
max,9.000000,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,254.000000,254.000000,253.000000,253.00000,254.000000,62.000000,0.0,0.0,0.0,0.0


### Finding the Label in the dataset

In [6]:
print(df.shape)
print(df.columns[:5])
print(df.columns[-5:])
print('label' in df.columns)

(42000, 785)
Index(['label', 'pixel0', 'pixel1', 'pixel2', 'pixel3'], dtype='object')
Index(['pixel779', 'pixel780', 'pixel781', 'pixel782', 'pixel783'], dtype='object')
True


## Checking for Missing Values

In [7]:
df.isna().sum().sum() # This dataset has no missing values due to the fact that it is a clean dataset.

np.int64(0)

## Removing those columns which are not required for the model training or are continously Zero

In [8]:
pixel_cols = [c for c in df.columns if c.startswith("pixel")]
all_zero_cols = [c for c in pixel_cols if (df[c] == 0).all()]
len(all_zero_cols), all_zero_cols[:10]
df = df.drop(columns=all_zero_cols)
print("Those columns which are continously zero and not required for model training have been removed.")

Those columns which are continously zero and not required for model training have been removed.


In [9]:
print(f"Dataset has {df.shape[0]} rows and {df.shape[1]} columns.")


Dataset has 42000 rows and 709 columns.


## Normalizing the features and separating the target variable from the features

In [10]:
pixel_cols = [c for c in df.columns if c.startswith("pixel")]

X = df[pixel_cols].astype(np.float32)
y = df["label"]  # or whatever the target column is called

X = X / 255.0
print("Features (X) and target (y) have been separated and features have been normalized.")

Features (X) and target (y) have been separated and features have been normalized.


## Train/ Test Split

In [11]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

print("Train shape:", X_train.shape, "Test shape:", X_test.shape)

Train shape: (33600, 708) Test shape: (8400, 708)


## Evaluation Function 

In [12]:
def evaluate_model(model, X_test, y_test, name="model", show_report=True, show_cm=True):
    """
    Fits are done outside this function.
    This function evaluates on X_test/y_test and prints:
    - accuracy
    - classification report
    - confusion matrix
    """
    y_pred = model.predict(X_test)

    acc = accuracy_score(y_test, y_pred)
    print(f"\n{name} Results")
    print("-" * (len(name) + 8))
    print("Accuracy:", round(acc, 4))

    if show_report:
        print("\nClassification report:")
        print(classification_report(y_test, y_pred))

    if show_cm:
        print("Confusion matrix:")
        print(confusion_matrix(y_test, y_pred))

    return acc

# Model Training and Evaluation

# KNN 
**Idea:**  
“To classify something new, look at the most similar things you’ve seen before, and copy the most common label.”

**Simple example (movies):**  
You want to guess a movie genre for a new movie.
- You look for the **5 most similar movies** (similar length, similar amount of action, similar humor).
- If **3 out of 5** are “Action”, you predict **Action**.

**How this relates to MNIST digits:**  
For a new handwritten digit image, k‑NN compares it with many saved images and says:
- “Which saved images look most similar to this one?”
- Then it takes a vote: if most similar ones are “7”, it predicts **7**.

**Strengths (in simple words):**
- Easy to understand.
- Often works decently as a baseline.

**Weaknesses (in simple words):**
- Can be **slow**, because it has to compare the new image to lots of old images.


In [13]:
knn_pipe = Pipeline([
    ("scaler", StandardScaler(with_mean=False) if hasattr(X_train, "toarray") else StandardScaler()),
    ("model", KNeighborsClassifier(n_neighbors=5))
])

# SVM 
**Idea:**  
“Draw the cleanest boundary that separates categories.”

Imagine you have points on a sheet:
- Red points = “cats”
- Blue points = “dogs”  
SVM tries to draw a line (or boundary) that separates them with the best “gap” between groups.

**Simple example (two fruits):**
You want to separate apples and oranges using two measurements:
- sweetness
- sourness  
SVM tries to find the best boundary so that apples fall on one side and oranges on the other.

**How this relates to MNIST digits:**
Each image has 784 pixel values. SVM tries to learn rules that separate digits:
- “This pattern looks like 0”
- “That pattern looks like 3”
- etc.

**Strengths (in simple words):**
- Often very accurate for many problems.
- Good at separating classes when the pattern is clear.

**Weaknesses (in simple words):**
- Can be **slow** on large image datasets (lots of rows and many pixel columns).
- Images are complex, and classical methods can struggle to capture “shape” the way humans see it.


In [14]:
svm_pipe = Pipeline([
    ("scaler", StandardScaler(with_mean=False) if hasattr(X_train, "toarray") else StandardScaler()),
    ("model", LinearSVC(random_state=42))
])

# Naive Bayes
**Idea:**  
“Use many small clues to guess the most likely category.”

**Simple example (spam detection):**
- If an email contains “free” → that’s a clue for spam
- If it contains “winner” → another clue for spam
- If it contains “meeting” → clue for not spam  
Naive Bayes combines the clues and picks the most likely answer.

**How this relates to MNIST digits:**  
It treats each pixel as a tiny clue.
- Dark pixels in certain places might suggest “1”
- Dark pixels in other places might suggest “8”

**Strengths (in simple words):**
- Very fast.
- Works well when “clues” are clear and consistent.

**Weaknesses (in simple words):**
- Pixels in an image are connected (shapes), but Naive Bayes treats clues as more independent than they really are.
- So it often doesn’t do as well on raw images.


In [15]:
# MultinomialNB expects non-negative features.
# Since you already normalized to 0..1, DO NOT StandardScale it (would create negatives).
nb_pipe = Pipeline([
    ("model", MultinomialNB())
])

### Training and Evaluation of the models

In [16]:
knn_pipe.fit(X_train, y_train)
evaluate_model(knn_pipe, X_test, y_test, name="k-NN (k=5)")


k-NN (k=5) Results
------------------
Accuracy: 0.9375

Classification report:
              precision    recall  f1-score   support

           0       0.95      0.98      0.96       827
           1       0.95      1.00      0.97       937
           2       0.95      0.94      0.94       835
           3       0.92      0.93      0.93       870
           4       0.94      0.92      0.93       814
           5       0.91      0.91      0.91       759
           6       0.96      0.97      0.97       827
           7       0.94      0.93      0.93       880
           8       0.97      0.86      0.92       813
           9       0.90      0.92      0.91       838

    accuracy                           0.94      8400
   macro avg       0.94      0.94      0.94      8400
weighted avg       0.94      0.94      0.94      8400

Confusion matrix:
[[811   0   1   1   0   6   6   1   0   1]
 [  0 933   0   1   0   0   2   1   0   0]
 [  8   8 781  14   4   1   5   9   2   3]
 [  4   2  11 

0.9375

In [17]:
nb_pipe.fit(X_train, y_train)
evaluate_model(nb_pipe, X_test, y_test, name="Naive Bayes (MultinomialNB)")


Naive Bayes (MultinomialNB) Results
-----------------------------------
Accuracy: 0.8246

Classification report:
              precision    recall  f1-score   support

           0       0.93      0.92      0.93       827
           1       0.89      0.93      0.91       937
           2       0.87      0.85      0.86       835
           3       0.80      0.81      0.80       870
           4       0.84      0.72      0.77       814
           5       0.84      0.64      0.73       759
           6       0.88      0.92      0.90       827
           7       0.96      0.83      0.89       880
           8       0.64      0.77      0.70       813
           9       0.68      0.84      0.75       838

    accuracy                           0.82      8400
   macro avg       0.83      0.82      0.82      8400
weighted avg       0.83      0.82      0.83      8400

Confusion matrix:
[[762   1   4   3   2   8  16   0  28   3]
 [  0 874   3   7   0   4   5   0  43   1]
 [  8   9 706  20   9  

0.8246428571428571

In [ ]:
svm_pipe.fit(X_train, y_train)
evaluate_model(svm_pipe, X_test, y_test, name="Linear SVM")
# We have not run this due to lack of GPU and it would take alot on this big of a Dataset.

# Report for the models
## Model Evaluation Report (MNIST Digits – Classical ML)

### Dataset + Setup (what I did)
- **Task:** Identify handwritten digits (**0–9**) from images stored as pixel values.
- **Data format:** Each image is represented by **784 pixel columns** (28×28 flattened into one row).
- **Split:** Training and testing were separated so the models were tested on **unseen images**.
- **Evaluation method:** I measured overall **accuracy** and also checked per-digit performance (how well each digit was recognized).

---

## 1) k‑Nearest Neighbors (k‑NN, k=5)

### Overall result
- **Accuracy:** **0.9375** (≈ **93.75%**)

### What this means (simple)
k‑NN works by comparing a new digit image to the most similar images it has already seen and then **voting** on the label.  
The high accuracy shows that “similar-looking digits often share the same label,” so this method is naturally strong on MNIST.

### Per-digit highlights (from the report)
- Strong performance on most digits, especially:
  - **Digit 1:** recall ≈ **1.00** (it almost always finds 1 correctly)
  - **Digit 6:** very high precision/recall (very consistent recognition)
- A noticeable weakness:
  - **Digit 8:** recall ≈ **0.86**, meaning the model misses 8 more often than other digits (it likely confuses 8 with digits that have similar shapes like 3, 9, or 0 depending on writing style).

### Key takeaway
- **k‑NN gives strong accuracy on MNIST**, but it can be **slow** because it compares test images to many training images.

---

## 2) Naive Bayes (MultinomialNB)

### Overall result
- **Accuracy:** **0.8246** (≈ **82.46%**)

### What this means (simple)
Naive Bayes predicts using many small “clues” and tries to choose the most likely digit.  
On images, this approach is weaker because the digit is not just a collection of independent pixels — the **shape and structure** matter (pixels work together to form strokes).

### Per-digit highlights (from the report)
- Stronger digits:
  - **0:** precision/recall both around **0.92+**
  - **1:** strong performance around **0.90+**
  - **6:** strong recall around **0.92**
- Weaker digits:
  - **8:** precision ≈ **0.64**, meaning when it predicts “8”, it is wrong more often than for other digits.
  - **5:** recall ≈ **0.64**, meaning it misses many actual 5s.
  - **4:** recall ≈ **0.72**, also comparatively weaker.
  - **9:** precision ≈ **0.68**, meaning 9 is also confused more frequently.

### Key takeaway
- **Naive Bayes is fast**, but **less accurate on raw pixel images** because it doesn’t naturally handle “shape” the way humans do.

---

## Comparison Summary

| Model | Accuracy | Main Strength | Main Limitation |
|------|----------|---------------|-----------------|
| k‑NN (k=5) | **93.75%** | Very strong on image similarity | Can be slow at prediction time |
| Naive Bayes | **82.46%** | Very fast baseline | Struggles with digits that require understanding shape (like 8, 5, 9) |

---

## Final Conclusion
For this MNIST-style pixel dataset:
- **k‑NN performed significantly better than Naive Bayes** because handwritten digits are mainly recognized by **visual similarity**, which k‑NN directly uses.
- **Naive Bayes provides a useful quick baseline**, but its assumptions make it less suitable for image pixels where the meaning comes from the **overall structure** of the digit, not individual pixels alone.

This comparison supports the main learning goal: **classical machine learning can work on images**, but it also shows limitations when the model does not naturally capture the idea of *shape*.

## Now saving the Best Model 


In [18]:
import joblib
joblib.dump(knn_pipe, "knn_mnist_model.joblib")
print("k-NN model saved as knn_mnist_model.joblib")

k-NN model saved as knn_mnist_model.joblib
